src/Data/Interval.hs

```diff
@@ -661,6 +674,135 @@ instance forall r. (Real r, Fractional r) => Fractional (Interval r) where
       lb3 = minimumBy cmpLB xs
       xs = [recipLB (lowerBound' a), recipUB (upperBound' a)]
 
+instance (RealFrac r, Floating r) => Floating (Interval r) where
+  pi = singleton pi
+
+  exp = intersection (0 <..< PosInf) . mapMonotonic exp
+  log a = interval (logB (lowerBound' b)) (logB (upperBound' b))
+    where
+      b = intersection (0 <..< PosInf) a
+
+  sqrt = mapMonotonic sqrt . intersection (0 <=..< PosInf)
+
+  -- | If results do not form a connected interval, throws 'LossOfPrecision'.
```

Comment:
Personally, I prefer to compute an over-approximation (i.e. convex hull) instead of throwing error in this case and to provide such variants (throwing exceptions or returning `Maybe`) as different API if necessary. Because

* such behavior is convenient in applications like abstract interpretation,
* not throwing an exception (when the operation for the base type does not throw exception) makes it easier to use interval values instead of regular values (e.g., passing `Interval a` instead of real number type `a` for an existing function `f :: (RealFrac a, Floating a) => ..`),
* some existing methods already choosed to do so: e.g. `signum (-10 <=..<= 10)` currently returns `Finite (-1) <=..<= Finite 1` (i.e. convex hull of `{-1, 0, +1}`) instead of throwing an exception.

This may be just my personal preference, so I would like to know if there are use cases where it is preferable to make it an error.

Reply:
I think that over-approximation in general is a slippery slope, where one is under a constant temptation to give up and return `whole`. Allowing over-approximation opens a way to non-equivalent definitions of the same function, which are nevertheless equally "correct".

That said, defining functions over intervals as a convex hull over pointwise applications is OK in the sense that it determines a function uniquely. But being precise and throwing an error has its own benefits: this is a clear sign that I'm doing something wrong and should resort to `IntervalSet` or express my intentions more clearly by reaching to a specific variant. My experience is that otherwise in complex computations it is difficult to track down at which particular step an over-approximation has happenned. 

I do not feel very strongly about this matter. We may need to review my choice in #21 with regards to division by zero. On the other hand, while I'd argue that one should expect that `recip` may result in `DivideByZero` (because it does so for `Double` as well), receiving `LossOfPrecision` from a usually "total" function is certainly surprising. 

BTW we need to document the behaviour of `signum`, because it violates `signum x * abs x == x` law.

Reply:
Another interesting thing that I'm throwing exceptions from a `Safe` module. While one cannot do `error` or `undefined`, GHC does not mind `throw DivideByZero` / `throw LossOfPrecision`. Could it be a bug in GHC?.. This is a strong argument in favor of convex hulls: it is probably dishonest to trick GHC and downstream users.

Reply:
Thank you for the explanation.
I think the rationale is convincing.

> I do not feel very strongly about this matter.

Now that I understand the reason, I also don't have a strong opinion on which one it should be.

> We may need to review my choice in #21 with regards to division by zero.

I think that `recip` resulting in error is consistent with the "convex hull over pointwise applications" view if we define convex hull of a set containing `undefined` as `undefined`.

> On the other hand, while I'd argue that one should expect that `recip` may result in `DivideByZero` (because it does so for `Double` as well), receiving `LossOfPrecision` from a usually "total" function is certainly surprising.

Yes, that is a point.

> BTW we need to document the behaviour of `signum`, because it violates `signum x * abs x == x` law.

I agree that `signum` case is worth documenting, as it violates the law and also loses precision.
On the other hand, `Interval` violates many algebraic laws, e.g. `x - x = 0` (if `x` is `-1 <=..<= 1`, `x - x` is not `singleton 0` but `-2 <=..<= 2`), and documenting all of them might be bothersome. 

> Another interesting thing that I'm throwing exceptions from a `Safe` module. While one cannot do `error` or `undefined`, GHC does not mind `throw DivideByZero` / `throw LossOfPrecision`. 

I don't know much about Safe Haskell, but I didn't know that `error` and `undefined` are forbidden in Safe Haskell..

Reply:
> I don't know much about Safe Haskell, but I didn't know that `error` and `undefined` are forbidden in Safe Haskell.

Seems I got it totally wrong: it appears that `error` and `undefined` are not forbidden. Then at the very least I should replace `throw LossOfPrecision` with a self-explaining error message.